### PR TITLE
store state faster and stop on errors

### PR DESCRIPTION
**Pull request by nvdk**
initially just wanted better state management (storing state faster and stopping on errors), ended up refactoring a bit. no longer uses the ldes-consumer library now.

The consumeStream method uses an async promise executor which is considered a bad pattern, but I think I've handled possible errors and don't really see a better way of writing this at the moment.

### CODE REVIEW

File: consumer.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+  constructor ({ endpoint, datasetIri, ldesOptions }: ConsumerArgs) {
+    this.endpoint = endpoint;
+    this.client = newEngine();
+    const defaultOptions = {
+      representation: "Quads",
+      mimeType: "application/ld+json",
+      requestHeaders: extractEndpointHeadersFromEnv(LDES_ENDPOINT_HEADER_PREFIX),
+      emitMemberOnce: true,
+      disableSynchronization: false
```
Comment (nvdk) [inline consumer.ts]: whoops this should be true, we don't expect the service to keep running when it's done

**Comment by nvdk** [comment]
Tested on two feeds:
- https://ipdc.vlaanderen.be/id/conceptsnapshot?pageNumber=0 (finishes without error)
- https://marineregions.org/feed stops on error

File: consumer.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      versionOfPath: LDES_VERSION_OF_PATH,
+      timestampPath: LDES_TIMESTAMP_PATH
+    };
+  }
+
+  consumeStream () {
+    return new Promise(async (resolve, reject) => {
+      if (this.isRunning) reject(new Error("already running"));
+      this.isRunning = true;
+      const lastState = await fetchState(this.datasetIri);
```
Comment (nvdk) [inline consumer.ts]: should probably add a `.catch((e) => reject(e)` here

File: consumer.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  endpoint: string
+  ldesOptions?: ConfigurableLDESOptions;
+};
+
+export type Member = {
+  id: RDF.Term;
+  quads: RDF.Quad[];
+};
+
+enum PAUZE_REASONS {
```
Comment (elpoelma) [inline consumer.ts]: This should probably be pause :)

**Review by elpoelma — COMMENTED** [review]
It seems to work well in combination with the ipdc LDES feed.

**Review by elpoelma — CHANGES_REQUESTED** [review]
Seems to work well overall. I did notice that the service hangs up when the endpoint URL is malformed.

**Comment by MPParsley** [comment]
"@treecg/actor-init-ldes-client": "https://github.com/redpencilio/event-stream-client/releases/download/actor-init-ldes-client-v3.0.0-fix-pausing/treecg-actor-init-ldes-client-3.0.0.tgz",

causes a

```
node:internal/modules/cjs/loader:936
  throw err;
  ^

Error: Cannot find module '@treecg/actor-init-ldes-client'
Require stack:
- /usr/src/build/consumer.js
- /usr/src/build/app.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/usr/src/build/consumer.js:7:28)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/usr/src/build/consumer.js', '/usr/src/build/app.js' ]
}
```

File: app.ts [hunk 4]
```diff
@@ -78,54 +23,20 @@ const consumerJob = new CronJob(CRON_PATTERN, async () => {
...
-          }
-        },
-        async (state) => {
-          console.log("CONSUMER DONE");
-          await updateState(stream, state);
-          taskIsRunning = false;
-          // Shutdown process when running as a Job.
-          if (RUNONCE) {
-            console.log("Job is complete.");
-            process.exit();
```
Comment (MPParsley) [inline app.ts:125]: This breaks the RUNONCE option.

**Review by MPParsley — CHANGES_REQUESTED** [review]

**Comment by nvdk** [comment]
closing in favour of #36

**Pull request closed without merging**